plugins/soundsourcem4a/audiosourcem4a.cpp

```diff
@@ -0,0 +1,307 @@
+#else
+    m_hFile = MP4Read(fileName.toLocal8Bit().constData());
+#endif
+    if (MP4_INVALID_FILE_HANDLE == m_hFile) {
+        qWarning() << "Failed to open file for reading:" << fileName;
+        return ERR;
+    }
+
+    m_trackId = findAacTrackId(m_hFile);
+    if (MP4_INVALID_TRACK_ID == m_trackId) {
+        qWarning() << "No AAC track found in file:" << fileName;
+        return ERR;
+    }
+
+    m_maxSampleId = MP4GetTrackNumberOfSamples(m_hFile, m_trackId);
+    if (MP4_INVALID_SAMPLE_ID == m_maxSampleId) {
+        qWarning() << "Failed to read file structure:" << fileName;
+        return ERR;
+    }
+    m_curSampleId = MP4_INVALID_SAMPLE_ID;
+
+    m_inputBuffer.resize(MP4GetTrackMaxSampleSize(m_hFile, m_trackId), 0);
+    m_inputBufferOffset = 0;
+    m_inputBufferLength = 0;
+
+    if (!m_hDecoder) {
+        // lazy initialization of the AAC decoder
+        m_hDecoder = faacDecOpen();
+        if (!m_hDecoder) {
+            qWarning() << "Failed to open the AAC decoder!";
+            return ERR;
+        }
+        faacDecConfigurationPtr pDecoderConfig = faacDecGetCurrentConfiguration(
+                m_hDecoder);
+        pDecoderConfig->outputFormat = FAAD_FMT_FLOAT; /* 32-bit float */
+        pDecoderConfig->downMatrix = 1; /* 5.1 -> stereo */
+        pDecoderConfig->defObjectType = LC;
+        if (!faacDecSetConfiguration(m_hDecoder, pDecoderConfig)) {
+            qWarning() << "Failed to configure AAC decoder!";
+            return ERR;
+        }
+    }
+
+    u_int8_t* configBuffer = NULL;
+    u_int32_t configBufferSize = 0;
+    if (!MP4GetTrackESConfiguration(m_hFile, m_trackId, &configBuffer,
+            &configBufferSize)) {
+        /* failed to get mpeg-4 audio config... this is ok.
+         * faacDecInit2() will simply use default values instead.
+         */
+        qWarning()
+                << "Failed to read the MP4 audio configuration. Continuing with default values.";
+    }
+
+    SAMPLERATE_TYPE sampleRate;
+    unsigned char channelCount;
+    if (0 > faacDecInit2(m_hDecoder, configBuffer, configBufferSize,
+                    &sampleRate, &channelCount)) {
+        free(configBuffer);
+        qWarning() << "Failed to initialize the AAC decoder!";
+        return ERR;
+    } else {
+        free(configBuffer);
+    }
+
+    setChannelCount(channelCount);
+    setFrameRate(sampleRate);
+    setFrameCount(m_maxSampleId * kFramesPerSampleBlock);
+
+    const SampleBuffer::size_type prefetchSampleBufferSize = (kSampleIdPrefetchCount + 1) * frames2samples(kFramesPerSampleBlock);
+    SampleBuffer(prefetchSampleBufferSize).swap(m_prefetchSampleBuffer);
+
+    // invalidate current frame index
+    m_curFrameIndex = getFrameCount();
+    // seek to beginning of file
+    if (0 != seekFrame(0)) {
+        qWarning() << "Failed to seek to the beginning of the file!";
+        return ERR;
+    }
+
+    return OK;
+}
+
+void AudioSourceM4A::close() throw() {
```

Comment:
This means throw() consumes all exceptions?

Thanks for explaining. The current design works for my, we have a corresponding open and close function. That is fine. An other design is to pass the audio file to the constructor and merge open and close with the constructor and destructor.  Not sure if it is worth the work.